server/aesdsocket.c

```diff
@@ -0,0 +1,201 @@
+
+    // Create a socket
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd == -1) {
+        syslog(LOG_ERR, "ERROR: Failed to create socket");
+        cleanup(EXIT_FAILURE);
+    }
+
+    // Allow for reuse of port 9000
+    int enable_reuse = 1; // Set to 1 to enable reuse of port 9000
+    if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &enable_reuse, sizeof(int)) == -1) {
+        syslog(LOG_ERR, "ERROR: Failed to setsockopt");
+        cleanup(EXIT_FAILURE);
+    }
+
+    // Bind to port 9000
+    struct sockaddr_in server_addr;
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    server_addr.sin_port = htons(9000);
+
+    if (bind(sockfd, (struct sockaddr*)&server_addr, sizeof(server_addr)) == -1) {
+        syslog(LOG_ERR, "ERROR: Failed to bind");
+        cleanup(EXIT_FAILURE);
+    }
+
+    close(-1);
+    // Listen for connections
+    if (listen(sockfd, 5) == -1) {
+        syslog(LOG_ERR, "ERROR: Failed to listen");
+        close(sockfd);
+        return -1;
+    }
+
+    // Open file for aesdsocketdata
+    char *aesddata_file = "/var/tmp/aesdsocketdata";
+    datafd = open(aesddata_file, O_CREAT | O_RDWR | O_APPEND, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+    if (datafd == -1){
+        syslog(LOG_ERR, "ERROR: Failed to create file - %s", aesddata_file);
+        closelog();
+        exit(EXIT_FAILURE);
+    }
+
+    // Accept connections in a loop
+    while (1) {
+        struct sockaddr_in client_addr;
+        socklen_t client_addr_len = sizeof(client_addr);
+        client_sockfd = accept(sockfd, (struct sockaddr*)&client_addr, &client_addr_len);
+        if (client_sockfd == -1) {
+            syslog(LOG_WARNING, "WARNING: Failed to accept, retrying ...");
+            continue; // Continue accepting connections
+        }
+
+        // Log accepted connection
+        char client_ip[INET_ADDRSTRLEN];
+        inet_ntop(AF_INET, &(client_addr.sin_addr), client_ip, INET_ADDRSTRLEN);
+        syslog(LOG_INFO, "Accepted connection from %s", client_ip);
+
+        // Receive and process data
+        size_t buffer_size = 1024;
+        char* buffer = (char *)malloc(buffer_size * sizeof(char));
+        if (buffer == NULL) {
+            syslog(LOG_ERR, "ERROR: Failed to malloc");
+            cleanup(EXIT_FAILURE);
+        }
+        memset(buffer, 0, buffer_size * sizeof(char));
+        ssize_t recv_size;
+        while ((recv_size = recv(client_sockfd, buffer, buffer_size, 0)) > 0) {
+            // Append data to file
+            if (write(datafd, buffer, recv_size) == -1) {
+                syslog(LOG_ERR, "ERROR: Failed to write to %s file", aesddata_file);
+                cleanup(EXIT_FAILURE);
+            }
+
+            // Send data back to client if a complete packet is received (ends with newline)
+            if (memchr(buffer, '\n', buffer_size) != NULL) {
```

Comment:
@mubeena12 what will happen if the size of a received packet is > `buffer_size` with this design?  Will this meet the assignment requirements?

Reply:
@dwalkes The packet is received in chunks of 1024 bytes (equal to `buffer_size`) within a while loop (line 170). So, I only need to check for the packet termination `\n` in our case within the current chunk. If the packet is > `buffer_size`, it will be received as multiple chunks, with each chunk being checked for `\n` termination.